src/transformations/defmodel.cpp

```diff
@@ -0,0 +1,445 @@
+        bool ret = (type == PJ_TYPE_GEOGRAPHIC_2D_CRS ||
+                    type == PJ_TYPE_GEOGRAPHIC_3D_CRS);
+        proj_destroy(P);
+        return ret;
+    }
+
+    void geographicToGeocentric(double lam, double phi, double height, double a,
+                                double b, double /*es*/, double &X, double &Y,
+                                double &Z) {
+        (void)a;
+        (void)b;
+        assert(cart->a == a);
+        assert(cart->b == b);
+        PJ_LPZ lpz;
+        lpz.lam = lam;
+        lpz.phi = phi;
+        lpz.z = height;
+        PJ_XYZ xyz = cart->fwd3d(lpz, cart);
+        X = xyz.x;
+        Y = xyz.y;
+        Z = xyz.z;
+    }
+
+    void geocentricToGeographic(double X, double Y, double Z, double a,
+                                double b, double /*es*/, double &lam,
+                                double &phi, double &height) {
+        (void)a;
+        (void)b;
+        assert(cart->a == a);
+        assert(cart->b == b);
+        PJ_XYZ xyz;
+        xyz.x = X;
+        xyz.y = Y;
+        xyz.z = Z;
+        PJ_LPZ lpz = cart->inv3d(xyz, cart);
+        lam = lpz.lam;
+        phi = lpz.phi;
+        height = lpz.z;
+    }
+
+#ifdef DEBUG_DEFMODEL
+    void log(const std::string &msg) {
+        pj_log(ctx, PJ_LOG_TRACE, "%s", msg.c_str());
+    }
+#endif
+
+  private:
+    EvaluatorIface(const EvaluatorIface &) = delete;
+    EvaluatorIface &operator=(const EvaluatorIface &) = delete;
+};
+
+struct defmodelData {
+    std::unique_ptr<Evaluator<Grid, GridSet, EvaluatorIface>> evaluator{};
+    EvaluatorIface evaluatorIface;
+
+    explicit defmodelData(PJ_CONTEXT *ctx, PJ *cart)
+        : evaluatorIface(ctx, cart) {}
+
+    defmodelData(const defmodelData &) = delete;
+    defmodelData &operator=(const defmodelData &) = delete;
+};
+
+} // namespace
+
+static PJ *destructor(PJ *P, int errlev) {
+    if (nullptr == P)
+        return nullptr;
+
+    auto Q = static_cast<struct defmodelData *>(P->opaque);
+    delete Q;
+    P->opaque = nullptr;
+
+    return pj_default_destructor(P, errlev);
+}
+
+static PJ_COORD forward_4d(PJ_COORD in, PJ *P) {
+    auto *Q = (struct defmodelData *)P->opaque;
+
+    PJ_COORD out;
+    out.xyzt.t = in.xyzt.t;
+
+    if (!Q->evaluator->forward(Q->evaluatorIface, in.xyzt.x, in.xyzt.y,
+                               in.xyzt.z, in.xyzt.t, out.xyzt.x, out.xyzt.y,
+                               out.xyzt.z)) {
+        return proj_coord_error();
+    }
+
+    return out;
+}
+
+static PJ_COORD reverse_4d(PJ_COORD in, PJ *P) {
+    auto *Q = (struct defmodelData *)P->opaque;
+
+    PJ_COORD out;
+    out.xyzt.t = in.xyzt.t;
+
+    if (!Q->evaluator->inverse(Q->evaluatorIface, in.xyzt.x, in.xyzt.y,
+                               in.xyzt.z, in.xyzt.t, out.xyzt.x, out.xyzt.y,
+                               out.xyzt.z)) {
+        return proj_coord_error();
+    }
+
+    return out;
+}
+
+static void reassign_context(PJ *P, PJ_CONTEXT *ctx) {
```

Comment:
These freestanding functions seem a little haphazard. If they were more specifically named with some comment on what they're for, I wouldn't ignore them at my peril as a reader 😄

Reply:
comment added